app/controllers/triannon/application_controller.rb

```diff
@@ -1,4 +1,55 @@
 module Triannon
   class ApplicationController < ActionController::Base
+
+
+    #--- Authentication methods
+    #
+    # The #access_token_data method is generally available to the
+    # application.  The #access_token_generate and #access_token_validate?
+    # methods are consolidated here to provide a unified view of these methods,
+    # although the application generally may not need to call them.  They are
+    # used specifically in the auth_controller and these methods are tested
+    # in the auth_controller_spec.
+
+    # construct and encrypt an access token, using login data
+    # save the token into session[:access_token]
+    def access_token_generate(data)
+      timestamp = Time.now.to_i.to_s # seconds since epoch
+      token = "#{data.to_json};;;#{timestamp}"
+      salt  = SecureRandom.random_bytes(64)
+      key   = ActiveSupport::KeyGenerator.new(timestamp).generate_key(salt)
+      crypt = ActiveSupport::MessageEncryptor.new(key)
+      session[:access_key] = key
+      session[:access_token] = crypt.encrypt_and_sign(token)
+    end
+
+    # decrypt, parse and validate access token
+    def access_token_valid?(code)
+      if code == session[:access_token]
+        key = session[:access_key]
+        crypt = ActiveSupport::MessageEncryptor.new(key)
+        token = crypt.decrypt_and_verify(code)
+        token_data = token.split(';;;')
+        identity = JSON.parse(token_data.first)
+        timestamp = token_data.last.to_i
+        elapsed = Time.now.to_i - timestamp  # sec since token was issued
```

Comment:
Thanks, that is interesting also, but the integer comparison is working as expected.  From my reading of time comparisons in ruby, it's best to use an integer offset from the epoch.  The use of 'hours.ago' may have round-off complications.